src/base.ts

```diff
@@ -0,0 +1,138 @@
+/* eslint no-underscore-dangle: ["error", { "allow": ["_config", "_tokenId", "_tokenSecret"] }] */
+
+import Axios, { AxiosInstance } from 'axios';
+import EventEmitter from 'events';
+import { RequestOptions } from './RequestOptions';
+
+const pkg = require('../package.json');
+
+/**
+ * Mux Base Class - Simple base class to be extended by all child modules.
+ *
+ * @ignore
+ * @property {string} tokenId - The ID for the access token.
+ * @property {string} tokenSecret - The secret for the access token.
+ * @property {object} config - The configuration for the Base object.
+ * @property {Object} requestOptions - The HTTP request options for Mux Assets
+ * @property {string} requestOptions.auth.username - HTTP basic auth username (access token)
+ * @property {string} requestOptions.auth.password - HTTP basic auth password (secret)
+ *
+ */
+export class Base extends EventEmitter {
+  readonly http: AxiosInstance;
+
+  private _tokenId!: string;
+  private _tokenSecret!: string;
+  private _config!: RequestOptions;
+
+  constructor(muxBase: Base);
+  constructor(requestOptions: RequestOptions);
+  constructor(tokenId: string, tokenSecret: string, config: RequestOptions);
+  constructor(
+    tokenIdOrOptionsOrBase: string | RequestOptions | Base,
+    tokenSecret?: string,
+    config?: RequestOptions
+  ) {
+    super();
+
+    if (tokenIdOrOptionsOrBase instanceof Base) {
+      // we could do this with Object.assign but I'd rather we be really explicit about what we copy.
+      this.config = tokenIdOrOptionsOrBase._config;
+      this._tokenId = tokenIdOrOptionsOrBase._tokenId;
+      this._tokenSecret = tokenIdOrOptionsOrBase._tokenSecret;
+
+      this.http = tokenIdOrOptionsOrBase.http;
+    } else {
+      if (
+        typeof tokenIdOrOptionsOrBase === 'object' &&
+        !(tokenIdOrOptionsOrBase instanceof Base)
```

Comment:
This is just being overly safe, right? It shouldn't actually make it here given the first evaluation

Reply:
Correct - this is a firewall check